src/renderer/components/NodeGridElement.tsx

```diff
@@ -71,15 +82,17 @@ const NodeGridElement: React.FC<NodeGridElementProps> = ({ node, graphName, temp
         }
     };
 
-    let externalPipes: PipeSegment[] = [];
-
-    if (connectedEth !== null) {
-        externalPipes = node
+    // TODO: pre-calculate external pipes
+    const externalPipes = useMemo(() => {
```

Comment:
this should probably move into computenode class to avoid calculations on render